expr/functions.go

```diff
@@ -58,407 +70,486 @@ var allFns = map[string]struct {
 	"Time.trunc":            {2, 2, timeTrunc},
 }
 
-func err(fn string, err error) (zngnative.Value, error) {
-	return zngnative.Value{}, fmt.Errorf("%s: %w", fn, err)
+//XXX this should be renamed so as not to clash with the conventional
+// id "err" used for the local variable
+func err(fn string, err error) (zng.Value, error) {
+	return zng.Value{}, fmt.Errorf("%s: %w", fn, err)
 }
 
-func lenFn(args []zngnative.Value) (zngnative.Value, error) {
-	switch zng.AliasedType(args[0].Type).(type) {
+func lenFn(args *Args) (zng.Value, error) {
+	switch zng.AliasedType(args.vals[0].Type).(type) {
 	case *zng.TypeArray, *zng.TypeSet:
-		v, err := args[0].ToZngValue()
+		v := args.vals[0]
+		len, err := v.ContainerLength()
 		if err != nil {
-			return zngnative.Value{}, err
+			return zng.Value{}, err
 		}
-		l, err := v.ContainerLength()
-		if err != nil {
-			return zngnative.Value{}, err
-		}
-		return zngnative.Value{zng.TypeInt64, int64(l)}, nil
+		return zng.Value{zng.TypeInt64, args.Int(int64(len))}, nil
```

Comment:
I see we're using the scratch for the generated `zcode.Bytes`, are we worried about GC overhead for all these `zng.Value` structs?

Reply:
No problems here as discussed.